articles/budougumi0617.re

```diff
@@ -1,18 +1,282 @@
+
+
+=== @<code>{golang.org/x/arch}パッケージ
+ * https://godoc.org/golang.org/x/arch
+ * https://github.com/golang/arch
+
+マシンアーキテクチャ依存のコードが同梱されている。Goのツールチェインで使われている。
+
+=== @<code>{golang.org/x/benchmarks}パッケージ
+ * https://godoc.org/golang.org/x/benchmarks
+ * https://github.com/golang/benchmarks
+
+
+ベンチマークを実行するツールがいくつか含まれているリポジトリ。
+ベンチマークの流れはドライバーで実装されている。ツール作る際の共通化の参考になりそう
+https://github.com/golang/benchmarks/blob/master//driver/driver.go
+
+GCについての知見？
+https://twitter.com/budougumi0617/status/1163561278335406080
+
+=== @<code>{golang.org/x/blog}パッケージ
+ * https://godoc.org/golang.org/x/blog
+ * https://github.com/golang/blog
+
+Go Blogのソースが置かれているリポジトリ。
+正直Go Blogは検索性があまり高くない。ブログ内容を検索したいとき、手元でgrepしたり、GitHub上で検索できる。
+ * https://blog.golang.org
+
+=== @<code>{golang.org/x/build}パッケージ
+ * https://godoc.org/golang.org/x/build
+ * https://github.com/golang/build
+
+=== @<code>{golang.org/x/crypto}パッケージ
+ * https://godoc.org/golang.org/x/crypto
+ * https://github.com/golang/crypto
+
+標準パッケージに含まれていない暗号形式の実装が含まれている。
+暗号化は自分で実装せず提供されているライブラリを使いましょう。
+
+=== @<code>{golang.org/x/debug}パッケージ
+ * @<href>{https://godoc.org/golang.org/x/debug}
+ * @<href>{https://github.com/golang/debug}
+
+
+debugツールであるviewcoreコマンドの実装が入っている。
+コアファイルの解析方法を知りたいときに役立つのかもしれない。
+正直delveで良いと思う。
+
+viewcoreコマンドの挙動は以下のように試すことができる。
+macOS Mojaveではうまくコアダンプファイルを作れなかったのでDockerでLinux環境を用意することをオススメします。
+//list[clash][無限ループで完了しないコード]{
+package main
+
+import "fmt"
+
+func goroutineA(c1 chan int, c2 chan int) {
+    _ = <-c2
+    c1 <- 1
+    return
+}
+
+func goroutineB(c1 chan int, c2 chan int) {
+    fmt.Println("c1", <-c1)
+    c2 <- 2
+    return
+}
+
+func main() {
+    c1 := make(chan int)
+    c2 := make(chan int)
+
+    go goroutineA(c1,c2)
+    go goroutineB(c1,c2)
+
+    for {
+    }
+}
+//}
+
+//cmd{
+$ docker run --rm --name sandbox -it golang:1.12.9-buster /bin/bash
+root@d5d112c8c215:/go/src# go get golang.org/x/debug/cmd/viewcore
+root@d5d112c8c215:/go/src# go build -gcflags "-N -l" main.go
+root@d5d112c8c215:/go/src# GOTRACEBACK=crash ./main &
+[1] 786
+root@d5d112c8c215:/go/src# kill -ABRT 786
+root@d5d112c8c215:/go/src# SIGABRT: abort
+PC=0x48739a m=0 sigcode=0
+
+goroutine 1 [running]:
+main.main()
+	/go/src/main.go:22 +0xba fp=0xc000074f98 sp=0xc000074f58 pc=0x48739a
+...
+
+[1]+  Aborted                 (core dumped) GOTRACEBACK=crash ./main
+root@d5d112c8c215:/go/src# viewcore core goroutines
+G stacksize=800
+  000000c000009918 000000c000009938 runtime.usleep+61
+  000000c000009938 000000c0000099b8 runtime.sighandler+976
+  000000c0000099b8 000000c000009a68 runtime.sigtrampgo+498
+  000000c00002c758 000000c00002c798 main.main+190
+  000000c00002c798 000000c00002c7e0 runtime.main+524
+...
+//}
+
+
+コアファイルの取得方法はここ。
+https://golang.org/doc/gdb
+
+=== @<code>{golang.org/x/exp}パッケージ
+https://godoc.org/golang.org/x/exp
+https://github.com/golang/exp
+
+試験的に実装されたコードが内包されています。
+例を挙げると、x/xerrorsも元はこのリポジトリで実装されていました。
+定期的にチェックしておくと、Goの新しい機能を先取りできるかも？
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "？". (ja-technical-writing/no-exclamation-question-mark)